pkg/manager/report_generator.go

```diff
@@ -0,0 +1,62 @@
+// Copyright 2015 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package manager
+
+import (
+	"sync"
+
+	"github.com/google/syzkaller/pkg/cover"
+	"github.com/google/syzkaller/pkg/cover/backend"
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/pkg/mgrconfig"
+	"github.com/google/syzkaller/pkg/vminfo"
+)
+
+type ReportGeneratorWrapper struct {
+	cfg     *mgrconfig.Config
+	modules []*vminfo.KernelModule
+
+	mu     sync.Mutex
+	cached *cover.ReportGenerator
+}
+
+func ReportGeneratorCache(cfg *mgrconfig.Config) *ReportGeneratorWrapper {
+	return &ReportGeneratorWrapper{cfg: cfg}
+}
+
+func (w *ReportGeneratorWrapper) Get() (*cover.ReportGenerator, error) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	if w.cached == nil {
+		log.Logf(0, "initializing coverage information...")
+		rg, err := cover.MakeReportGenerator(w.cfg, w.cfg.KernelSubsystem, w.modules, w.cfg.RawCover)
```

Comment:
does it make sense to create generator while we don't have modules?
it looks wrong, we will do lots of work that we will throw away and the generator will produce wrong data

Reply:
it's better to return "not initialized" error

Reply:
Are we guaranteed to always have non-empty modules array?

Reply:
I don't remember if we add some stub entry for main kernel or not.

Reply:
But probably better to not rely on that.

Reply:
Updated the code